source/module_io/write_dmr.cpp

```diff
@@ -0,0 +1,93 @@
+#include "write_dmr.h"
+
+#include "module_hamilt_lcao/module_hcontainer/hcontainer_funcs.h"
+#include "module_hamilt_lcao/module_hcontainer/output_hcontainer.h"
+#include "module_hamilt_pw/hamilt_pwdft/global.h"
+
+#include <iostream>
+
+namespace ModuleIO
+{
+std::string dmr_gen_fname(const int out_type, const int ispin, const bool append, const int istep)
+{
+    std::string fname = "dmr.csr";
+    if (out_type == 1)
+    {
+        if (!append && istep >= 0)
+        {
+            fname = std::to_string(istep + 1) + "_data-DMR-sparse_SPIN" + std::to_string(ispin) + ".csr";
+        }
+        else
+        {
+            fname = "data-DMR-sparse_SPIN" + std::to_string(ispin) + ".csr";
+        }
+    }
+    else if (out_type == 2)
+    {
+        fname = "output_DM" + std::to_string(ispin) + ".npz";
+    }
+    else
+    {
+        ModuleBase::WARNING("write_dmr", "the output type of DMR should be npz or csr.");
+    }
+    return fname;
+}
+
+void write_dmr_csr(std::string& fname, hamilt::HContainer<double>* dm_serial, const int istep)
+{
+    // write the head: ION step number, basis number and R loop number
+    std::ofstream ofs(fname, std::ios::app);
+    ofs << "STEP: " << istep << std::endl;
+    ofs << "Matrix Dimension of DM(R): " << GlobalV::NLOCAL << std::endl;
+    ofs << "Matrix number of DM(R): " << dm_serial->size_R_loop() << std::endl;
+
+    // write HR_serial to ofs
+    double sparse_threshold = 1e-10;
+    int precision = 8;
+    hamilt::Output_HContainer<double> out_dm(dm_serial, GlobalV::NLOCAL, GlobalV::NLOCAL, ofs, sparse_threshold, precision);
```

Comment:
do not use GlobalV::NLOCAL,

Reply:
Here, this function needs the dimension of H matrix. I add a new input parameter `nbasis` in function `write_dmr()` to avoid using GlobalV::NLOCAL in function. 
The value of `nbasis` should be get from class HContainer, and after `get_nabasis()` implemented in class HContainer, the input parameter `nbasis` can be removed from `write_dmr()`.